src/main/java/io/cdap/plugin/MultiTableDBSource.java

```diff
@@ -158,14 +163,29 @@ public void setContextForMultiTableDBInput(BatchSourceContext context,
 
     SettableArguments arguments = context.getArguments();
     for (DBTableInfo tableInfo : tables) {
+      Schema schema = tableInfo.getSchema();
       arguments.set(DynamicMultiFilesetSink.TABLE_PREFIX + tableInfo.getDbTableName().getTable(),
-                    tableInfo.getSchema().toString());
+                    schema.toString());
+      emitLineage(context, tableInfo, schema);
     }
 
     context.setInput(Input.of(conf.getReferenceName(),
                               new SourceInputFormatProvider(ErrorCollectingMultiTableDBInputFormat.class, hConf)));
   }
 
+  private void emitLineage(BatchSourceContext context, DBTableInfo tableInfo, Schema schema) {
+    Asset asset = Asset.builder(conf.getReferenceName())
+      .setFqn(FQNGenerator.constructFQN(conf.getConnectionString(), tableInfo.getDbTableName().getTable()))
+      .setTag(tableInfo.getDbTableName().getTable()).build();
+    LineageRecorder lineageRecorder = new LineageRecorder(context, asset);
+    lineageRecorder.createExternalDataset(schema);
+    if (schema != null && schema.getFields() != null) {
+      String operationName = "Read_from_" + tableInfo.getDbTableName().getTable();
```

Comment:
Check if operation names have character limitations

Reply:
Could not find any validations for operation name

Reply:
you just need to make sure operation name is unique per stage